### ISSUE TITLE
Allow configuration of ports on the frontend (nginx) container.

### ISSUE THREAD

**Issue by golyalpha** [caprover/caprover#666]
## Problem - why should the feature be added
Currently, in more complex setups, where, for example, there are already containers running, CapRover won't be able to start it's nginx container, as it currently attempts to bind to the standard 443 and 80 ports, which will fail, if for example. there's a HAProxy service running on the server already, responsible for routing request to backends based on domain (as is my case currently)

## Solution - added feature
Allow the user to change ports the nginx container binds to on the host machine.

## Workaround
Directly change the service spec using the CLI.

1. Scale captain-nginx service to 0 `docker service scale captain-nginx=0`
2. Update the service to use different ports (replace published=8080 and published=8443 as needed)
 `docker service update --publish-rm published=80,target=80,mode=host --publish-rm published=443,target=443,mode=host --publish-add published=8080,target=80,mode=host --publish-add published=8443,target=443,mode=host captain-nginx`
3. Scale captain-nginx service back to 1 `docker service scale captain-nginx=1`docker

As you can see, the command is relatively verbose, and I think there is a benefit to exposing said option in the UI and potentially even the CLI, at least for cases where the nginx service fails to launch for any reason.

**Comment by githubsaturn** [comment]
There is a config param that you can override if needed:
https://github.com/caprover/caprover/blob/52e0c2a8bfe9f525d69a411540f7e2e4697e0cfe/src/utils/CaptainConstants.ts#L140

Note that this is definitely not a case where we want to support in the GUI or as a standard case where we want to support. Certbot SSL generation works only with standard ports, hence issuing the SSL cert on a non standard port cannot be supported.

**Comment by golyalpha** [comment]
I disagree. The CapRover instance *can be* exposed on the standard ports, provided it's accessed using the domains assigned to it (HAProxy allows SSL Passthrough with SNI), in which case, everything works as expected. This is of course provided HAProxy (or another server which is capable of SNI with SSL Passthrough, rather than SSL Termination) is properly configured  (not sure about Apache, but I wasn't able to get SSL Passthrough setup working with NGINX).
'
After all, my setup does work with Captain NGINX using 8081 and 8444 with HAProxy bound to the standard 80 and 443 ports.

**Comment by golyalpha** [comment]
Here's the HAProxy configuration that makes this possible:
```
frontend http
        bind *:80
        use_backend caprover if { req.hdr(Host) -i -m end caprover-host.com }

frontend https
        mode tcp
        bind *:443
        option tcplog
        tcp-request inspect-delay 5s
        tcp-request content accept if { req_ssl_hello_type 1 }

        use_backend caprover-ssl if { req.ssl_sni -i -m end caprover-host.com }

backend caprover
        mode http
        server master localhost:8081

backend caprover-ssl
        mode tcp
        option ssl-hello-chk
        server master localhost:8444
```

The setup described in the config above is rather pointless, as all it does is add unnecessary complexity to the setup, as currently, the only job HAProxy is performing is forwarding data from ports 80 and 443 to servers on ports 8081 and 8444, but it should be noted, that I've done this for the sake of readability, and in all other cases, you'd have other backends and routing rules defined for the other applications running on your network.

**Comment by githubsaturn** [comment]
There are a few points here:

1) You do not need to rebuild the image if you want to modify AppConstants.config parameters, you can create an override file that overrides these params, see [here](https://caprover.com/docs/troubleshooting.html#customize-config-settings).

2) Manually updating the service to change the ports is not going to be persistent. On every restart, CapRover resets the nginx service to its default settings. 

3) Certbot webserver challenge only works with the HTTP port being setup on port 80:
![image](https://user-images.githubusercontent.com/10582082/81478200-c4992600-91e9-11ea-8a8a-561de9f2c7d2.png)

Now that you put CapRover behind another load balancer, things get complex. Let's say CapRover "thinks" it's listening to port 8081 for HTTP requests, but in fact, it's behind a loadbalancer that passes through port 80 to 8081. In that case, all the healthchecks and verifications are going to fail. 

Basically what you want is a situation where CapRover thinks it's using 80/443, but in fact nginx is listening to 8081/8444 and HAProxy is managing the port forwarding. This is indeed a very special and highly customized use case. In this case, the only clean way I can think of is to run a VM and run CapRover container in a VM or in a DiD (docker in docker).

**Comment by golyalpha** [comment]
Well, I did propose letting us change the docker port bindings for the nginx container inside the CapRover WebUI to a) make it easier to make that change, and b) make it easier to make that change persistent. Since CapRover also seems to require to know where

If we just change the service port bindings, and none of the assumptions caprover makes about the entire system (like which port the HTTP and HTTPS services are on), there's no need to run a VM or a DinD setup to get this all to work.

I admit that either way, this is a little more advanced setup than the average user of CapRover will use, but at the same time, I believe that it should be supported.

**Comment by githubsaturn** [comment]
> I admit that either way, this is a little more advanced setup than the average user of CapRover will use, but at the same time, I believe that it should be supported.

"_It should be supported_" is subjective. Like I said, it all comes down to the maintainability vs usage you get from a feature. The goals and scope of project is defined in [here](https://github.com/caprover/caprover/blob/master/CONTRIBUTING.md#important-caprover-goals--scope):
> The goal is to make the common use-cases exposed via simple controls on UI while allowing further customizations to be done through hooks and scripts. If a new feature is doable via the existing features, or a basic tool, do not add it to CapRover. We do not want to bloat this application. One example is: "Add a flag to customize the placement constraints of containers". This is definitely doable in Docker, but we don't want to mirror every single functionality of Docker to CapRover. If we do that, CapRover becomes a very hard to maintain project.

**Comment by golyalpha** [comment]
Well, if there's a way I missed, that lets me change ports on the nginx service when CapRover creates it, while letting CapRover still think that it's on ports 80 and 443, I'd love to hear it, mainly because it would (hopefully) save me from writing shell scripts or having to log on to the server after reboots (which don't happen often, so I'll likely just forget about them).
